docs/architecture/adr-031-typed-events.md

```diff
@@ -0,0 +1,421 @@
+# ADR 030: Typed Events
+
+## Changelog
+
+- 28-Sept-2020: Initial Draft
+
+## Authors
+
+- Anil Kumar (@anilcse)
+- Jack Zampolin (@jackzampolin)
+- Adam Bozanich (@boz) 
+
+## Status
+
+Proposed
+
+## Context
+
+Currently in the SDK, events are not properly organized (event data is defined in the handlers for each message) and are prone to inconsistencies. This makes customizing events difficult. In addition, they are difficult to consume. This proposal focuses on updating the events to use **typed events** in each module such that emiting and subscribing to events will be much more egrnomic. This will make use of the SDK to build event driven processes much easier and enable rapid development of things like relayers and automated transaction bots for SDK based chains. These types of bots enable will enable easy building of new features at exchanges, wallets, explorers, and defi protocols. IBC especially will benefit from this proposal.
+
+The end of this proposal contains a detailed example of how to consume events after this refactor.
+
+## Decision
+
+__Step-1__: Declare event types for `sdk.Msg`s a module implements using the typed event interface: `sdk.ModuleEvent`. We first need to define this interface and supporting types.
+
+```go
+// types/events.go
+
+// ModuleEvent is the interface that all message events will implement
+type ModuleEvent interface {
+    Context()   BaseModuleEvent
+    ABCIEvent() Event
+}
+
+// BaseModuleEvent contains information about the Module and Action of an event
+type BaseModuleEvent struct {
+    Module string
+    Action string
+}
+```
+
+The `BaseModuleEvent` struct will be used for basic context about the event. It aids in routing events to their 
+`Module` and `Action` specific event parsers.
+
+__Step 2__:  Implement additional functionality in the `types` package: utility functions and a parser to route the event to its proper module.
+
+When we subscribe to emitted events on the tendermint websocket, they are emitted in the form of an `abci.Event`. The parser will process this event using `sdk.NewSDKEvent(abci.Event)` to enable passing of the processed event to the proper module.
+
+```go
+// types/events.go
+
+// SDKEvent contains the string representation of the event and the module information
+type SDKEvent struct {
+    Sev  StringEvent
```

Comment:
Why `Sev`? Also, you don't define what `StringEvent` is. Does `Sev` not encapsulate `Base` (seems this way based off of the godoc)?

Reply:
Maybe this is a typo, I'm trying to reference the `StringEvents` type already defined in the `events.go`. `StringEvents` does contain the `Module` and `Action` but it is inconvenient to pull out. We could also add some functions here to `StringEvents` (`func (sev StringEvents) Module() string` and `func (sev StringEvents) Action() string`) and dispense with the `BaseModuleEvent` and `SDKEvent` types.